apps/ex_wire/lib/ex_wire/packet/status.ex

```diff
@@ -89,43 +96,36 @@ defmodule ExWire.Packet.Status do
   end
 
   @doc """
-  Given an RLP-encoded Status packet from Eth Wire Protocol, decodes into a Status packet.
-
-  Note: we will decode warp's `manifest_hash` and `block_number`, if given.
+  Given an RLP-encoded Status packet from Eth Wire Protocol, decodes into a
+  Status packet.
 
   ## Examples
 
       iex> ExWire.Packet.Status.deserialize([<<0x63>>, <<3>>, <<10>>, <<5>>, <<4>>])
-      %ExWire.Packet.Status{protocol_version: 0x63, network_id: 3, total_difficulty: 10, best_hash: <<5>>, genesis_hash: <<4>>}
-
-      iex> ExWire.Packet.Status.deserialize([<<0x63>>, <<3>>, <<10>>, <<5>>, <<4>>, <<11>>, <<11>>])
-      %ExWire.Packet.Status{protocol_version: 0x63, network_id: 3, total_difficulty: 10, best_hash: <<5>>, genesis_hash: <<4>>, manifest_hash: <<11>>, block_number: 11}
+      %ExWire.Packet.Status{
+        protocol_version: 0x63,
+        network_id: 3,
+        total_difficulty: 10,
+        best_hash: <<5>>,
+        genesis_hash: <<4>>
+      }
   """
   @spec deserialize(ExRLP.t()) :: t
   def deserialize(rlp) do
-    [protocol_version | rlp_tail] = rlp
-    [network_id | rlp_tail] = rlp_tail
-    [total_difficulty | rlp_tail] = rlp_tail
-    [best_hash | rlp_tail] = rlp_tail
-    [genesis_hash | rest] = rlp_tail
-
-    {manifest_hash, block_number} =
-      case rest do
-        [] ->
-          {nil, nil}
-
-        [manifest_hash, block_number] ->
-          {manifest_hash, block_number |> :binary.decode_unsigned()}
-      end
+    [
+      protocol_version,
+      network_id,
+      total_difficulty,
+      best_hash,
+      genesis_hash
+    ] = rlp
 
     %__MODULE__{
-      protocol_version: protocol_version |> :binary.decode_unsigned(),
-      network_id: network_id |> :binary.decode_unsigned(),
-      total_difficulty: total_difficulty |> :binary.decode_unsigned(),
+      protocol_version: :binary.decode_unsigned(protocol_version),
```

Comment:
So RLP encoding encodes `<<5>>` and `5` the same way, so there's no way to tell if we wanted a binary or an integer. In ExRLP, we decode to return a binary, but then it bites us because a lot of our code expects `val == ExRLP.decode(ExRLP.encode(val))` to be true, but for the value `[5]`, the result is `[<<5>>]`. Then, this decode function tries to get deserialize that list value into a struct, but it's stuck not knowing if it's going to get `5` or `<<5>>`, and it's just been too much effort to control for the input when we can just accept either. This `Exth. maybe_decode_unsigned` takes either and returns `5`.

Reply:
I like the flexibility, but the fact that @MattMSumner needed to ask (and I was going to as well), suggests that the name could be improved. How do you feel about something like `decode_to_unsigned_scalar`?